S3 Transfer Manager host/endpoint mismatch

### Describe the bug

I'm using localstack and notice that changing the `endpointOverride` causes failures with transfer with message:
```
Cannot create meta s3 request; host header value a.b.c.d does not match endpoint override a.b.c.d:4566
```

### Expected Behavior

S3 Transfer Manager should work.

### Current Behavior

Library:
```
s3-transfer-manager v2.19.25
aws-crt v0.21.2
```


Error:
```
Exception in thread "main" java.util.concurrent.CompletionException: software.amazon.awssdk.core.exception.SdkClientException: Unable to execute HTTP request: S3Client.aws_s3_client_make_meta_request: creating aws_s3_meta_request failed (aws_last_error: AWS_ERROR_INVALID_ARGUMENT(34), An invalid argument was passed to a function.)
    at software.amazon.awssdk.utils.CompletableFutureUtils.errorAsCompletionException(CompletableFutureUtils.java:65)
...
    at software.amazon.awssdk.transfer.s3.S3TransferManager.upload(S3TransferManager.java:447)
    at org.example.TransferBug.main(TransferBug.java:31)
Caused by: software.amazon.awssdk.core.exception.SdkClientException: Unable to execute HTTP request: S3Client.aws_s3_client_make_meta_request: creating aws_s3_meta_request failed (aws_last_error: AWS_ERROR_INVALID_ARGUMENT(34), An invalid argument was passed to a function.)
    at software.amazon.awssdk.core.exception.SdkClientException$BuilderImpl.build(SdkClientException.java:111)
    at software.amazon.awssdk.core.exception.SdkClientException.create(SdkClientException.java:47)
...
    at software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncExecutionFailureExceptionReportingStage.execute(AsyncExecutionFailureExceptionReportingStage.java:41)
    ... 18 more
Caused by: software.amazon.awssdk.crt.CrtRuntimeException: S3Client.aws_s3_client_make_meta_request: creating aws_s3_meta_request failed (aws_last_error: AWS_ERROR_INVALID_ARGUMENT(34), An invalid argument was passed to a function.) AWS_ERROR_INVALID_ARGUMENT(34)
    at software.amazon.awssdk.crt.s3.S3Client.s3ClientMakeMetaRequest(Native Method)
    at software.amazon.awssdk.crt.s3.S3Client.makeMetaRequest(S3Client.java:79)
...
    at software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncRetryableStage$RetryingExecutor.attemptExecute(AsyncRetryableStage.java:144)
    ... 25 more
```

I've tried various endpoint, and got the same result:
```
[ERROR] [2023-01-27T03:21:10Z] [000000016b70b000] [S3Client] - id=0x121f0f010 Cannot create meta s3 request; host header value 127.0.0.1 does not match endpoint override 127.0.0.1:4566
```

```
[ERROR] [2023-01-27T03:21:27Z] [000000016d0e7000] [S3Client] - id=0x11ee09010 Cannot create meta s3 request; host header value fake-test-bucket.localhost does not match endpoint override fake-test-bucket.localhost:4566
```

```
[ERROR] [2023-01-27T03:21:49Z] [000000016b68f000] [S3Client] - id=0x142e0caa0 Cannot create meta s3 request; host header value fake-test-bucket.s3.localhost.localstack.cloud does not match endpoint override fake-test-bucket.s3.localhost.localstack.cloud:4566
```


### Reproduction Steps

Running the code on localstack throws the error.

```
package org.example;

import java.net.URI;

import software.amazon.awssdk.core.async.AsyncRequestBody;
import software.amazon.awssdk.crt.Log;
import software.amazon.awssdk.services.s3.S3AsyncClient;
import software.amazon.awssdk.services.s3.model.PutObjectRequest;
import software.amazon.awssdk.transfer.s3.S3TransferManager;

public class TransferBug {
    public static void main(String... args) {
        Log.initLoggingToFile(Log.LogLevel.Error, "file-output.txt");

        var endpoint = "http://s3.localhost.localstack.cloud:4566";
        var bucket = "fake-test-bucket";
        var key = "test.txt";

        var client = S3AsyncClient.crtBuilder()
                .endpointOverride(URI.create(endpoint))
                .build();

        try (var s3TransferManager = S3TransferManager.builder()
                .s3Client(client)
                .build()
        ) {
            var putRequest = PutObjectRequest.builder()
                    .bucket(bucket).key(key)
                    .build();

            var upload = s3TransferManager.upload(b -> b.putObjectRequest(putRequest)
                    .requestBody(AsyncRequestBody.fromString("test")));

            upload.completionFuture().join();
        }
    }
}
```


docker-compose.yaml
```
services:
  # https://github.com/localstack/localstack/blob/v1.3.1/docker-compose.yml
  localstack:
    image: localstack/localstack:1.3.1
    ports:
      - "127.0.0.1:4566:4566"            # LocalStack Gateway
      - "127.0.0.1:4510-4559:4510-4559"  # external services port range
    environment:
      - DEBUG=${DEBUG-}
      - LAMBDA_EXECUTOR=${LAMBDA_EXECUTOR-}
      - DOCKER_HOST=unix:///var/run/docker.sock
    volumes:
      - "${LOCALSTACK_VOLUME_DIR:-./volume}:/var/lib/localstack"
      - "/var/run/docker.sock:/var/run/docker.sock"
```

### Possible Solution

I downgraded `aws-crt` to `v0.21.1`, and it works for some endpoints.

* `s3.localhost.localstack.cloud` => works
* `127.0.0.1` => works
* `localhost` => fails, likely due to missing `pathStyleAccessEnabled`

The issue appears to be caused by https://github.com/awslabs/aws-crt-java/pull/581, and not sure if it is intentional.

### Additional Information/Context

_No response_

### AWS Java SDK version used

2.19.25

### JDK version used

openjdk version "17.0.4" 2022-07-19 LTS; OpenJDK Runtime Environment Corretto-17.0.4.8.1 (build 17.0.4+8-LTS)

### Operating System and version

MacOS 12.6 (Monterey)

--- a/.changes/next-release/feature-AWSCRTbasedS3Client-dcd8e15.json
+++ b/.changes/next-release/feature-AWSCRTbasedS3Client-dcd8e15.json
@@ -1,0 +1,6 @@
+{
+    "type": "bugfix",
+    "category": "AWS CRT-based S3 Client",
+    "contributor": "",
+    "description": "Fixed an issue where port number was not added to host header. See [#3721](https://github.com/aws/aws-sdk-java-v2/issues/3721) and [#3682](https://github.com/aws/aws-sdk-java-v2/issues/3682)."
+}

--- a/pom.xml
+++ b/pom.xml
@@ -114,7 +114,7 @@
         <rxjava.version>2.2.21</rxjava.version>
         <commons-codec.verion>1.10</commons-codec.verion>
         <jmh.version>1.29</jmh.version>
-        <awscrt.version>0.20.3</awscrt.version>
+        <awscrt.version>0.21.3</awscrt.version>
 
         <!--Test dependencies -->
         <junit5.version>5.8.1</junit5.version>

--- a/services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/crt/S3CrtAsyncHttpClient.java
+++ b/services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/crt/S3CrtAsyncHttpClient.java
@@ -44,6 +44,7 @@ import software.amazon.awssdk.http.async.AsyncExecuteRequest;
 import software.amazon.awssdk.http.async.SdkAsyncHttpClient;
 import software.amazon.awssdk.utils.AttributeMap;
 import software.amazon.awssdk.utils.Logger;
+import software.amazon.awssdk.utils.http.SdkHttpUtils;
 
 /**
  * An implementation of {@link SdkAsyncHttpClient} that uses an CRT S3 HTTP client {@link S3Client} to communicate with S3.
@@ -86,7 +87,7 @@ public final class S3CrtAsyncHttpClient implements SdkAsyncHttpClient {
     public CompletableFuture<Void> execute(AsyncExecuteRequest asyncRequest) {
         CompletableFuture<Void> executeFuture = new CompletableFuture<>();
         URI uri = asyncRequest.request().getUri();
-        HttpRequest httpRequest = toCrtRequest(uri, asyncRequest);
+        HttpRequest httpRequest = toCrtRequest(asyncRequest);
         S3CrtResponseHandlerAdapter responseHandler =
             new S3CrtResponseHandlerAdapter(executeFuture, asyncRequest.responseHandler());
 
@@ -97,12 +98,13 @@ public final class S3CrtAsyncHttpClient implements SdkAsyncHttpClient {
 
         ChecksumConfig checksumConfig =
             checksumConfig(httpChecksum, requestType, s3NativeClientConfiguration.checksumValidationEnabled());
+        URI endpoint = getEndpoint(uri);
 
         S3MetaRequestOptions requestOptions = new S3MetaRequestOptions()
             .withHttpRequest(httpRequest)
             .withMetaRequestType(requestType)
             .withChecksumConfig(checksumConfig)
-            .withEndpoint(getEndpoint(uri))
+            .withEndpoint(endpoint)
             .withResponseHandler(responseHandler)
             .withResumeToken(resumeToken);
 
@@ -156,7 +158,7 @@ public final class S3CrtAsyncHttpClient implements SdkAsyncHttpClient {
         });
     }
 
-    private static HttpRequest toCrtRequest(URI uri, AsyncExecuteRequest asyncRequest) {
+    private static HttpRequest toCrtRequest(AsyncExecuteRequest asyncRequest) {
         SdkHttpRequest sdkRequest = asyncRequest.request();
 
         String method = sdkRequest.method().name();
@@ -169,7 +171,7 @@ public final class S3CrtAsyncHttpClient implements SdkAsyncHttpClient {
                                               .map(value -> "?" + value)
                                               .orElse("");
 
-        HttpHeader[] crtHeaderArray = createHttpHeaderList(uri, asyncRequest).toArray(new HttpHeader[0]);
+        HttpHeader[] crtHeaderArray = createHttpHeaderList(asyncRequest).toArray(new HttpHeader[0]);
 
         S3CrtRequestBodyStreamAdapter sdkToCrtRequestPublisher =
             new S3CrtRequestBodyStreamAdapter(asyncRequest.requestContentPublisher());
@@ -207,13 +209,14 @@ public final class S3CrtAsyncHttpClient implements SdkAsyncHttpClient {
         }
     }
 
-    private static List<HttpHeader> createHttpHeaderList(URI uri, AsyncExecuteRequest asyncRequest) {
+    private static List<HttpHeader> createHttpHeaderList(AsyncExecuteRequest asyncRequest) {
         SdkHttpRequest sdkRequest = asyncRequest.request();
         List<HttpHeader> crtHeaderList = new ArrayList<>();
 
         // Set Host Header if needed
         if (!sdkRequest.firstMatchingHeader(Header.HOST).isPresent()) {
-            crtHeaderList.add(new HttpHeader(Header.HOST, uri.getHost()));
+            String hostHeader = getHostHeaderValue(asyncRequest.request());
+            crtHeaderList.add(new HttpHeader(Header.HOST, hostHeader));
         }
 
         // Set Content-Length if needed
@@ -227,5 +230,11 @@ public final class S3CrtAsyncHttpClient implements SdkAsyncHttpClient {
                                                       .forEach(crtHeaderList::add));
 
         return crtHeaderList;
+    }
+
+    private static String getHostHeaderValue(SdkHttpRequest request) {
+        return SdkHttpUtils.isUsingStandardPort(request.protocol(), request.port())
+               ? request.host()
+               : request.host() + ":" + request.port();
     }
 }

--- a/services/s3/src/test/java/software/amazon/awssdk/services/s3/functionaltests/GetBucketPolicyFunctionalTest.java
+++ b/services/s3/src/test/java/software/amazon/awssdk/services/s3/functionaltests/GetBucketPolicyFunctionalTest.java
@@ -27,15 +27,16 @@ import org.junit.Rule;
 import org.junit.Test;
 import software.amazon.awssdk.auth.credentials.AwsBasicCredentials;
 import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider;
+import software.amazon.awssdk.crt.Log;
 import software.amazon.awssdk.regions.Region;
 import software.amazon.awssdk.services.s3.S3AsyncClient;
 import software.amazon.awssdk.services.s3.S3AsyncClientBuilder;
 import software.amazon.awssdk.services.s3.S3Client;
 import software.amazon.awssdk.services.s3.S3ClientBuilder;
+import software.amazon.awssdk.services.s3.S3CrtAsyncClientBuilder;
 import software.amazon.awssdk.services.s3.model.GetBucketPolicyResponse;
 
 public class GetBucketPolicyFunctionalTest {
-    private static final URI HTTP_LOCALHOST_URI = URI.create("http://localhost:8080/");
     private static final String EXAMPLE_BUCKET = "Example-Bucket";
     private static final String EXAMPLE_POLICY =
         "{\"Version\":\"2012-10-17\",\"Id\":\"Policy1234\","
@@ -44,13 +45,17 @@ public class GetBucketPolicyFunctionalTest {
         + "\"Resource\":\"arn:aws:s3:::dummy-resource/*\"}]}";
 
     @Rule
-    public WireMockRule wireMock = new WireMockRule();
+    public WireMockRule wireMock = new WireMockRule(0);
+
+    private URI endpoint() {
+        return URI.create("http://localhost:" + wireMock.port());
+    }
 
     private S3ClientBuilder getSyncClientBuilder() {
 
         return S3Client.builder()
                        .region(Region.US_EAST_1)
-                       .endpointOverride(HTTP_LOCALHOST_URI)
+                       .endpointOverride(endpoint())
                        .credentialsProvider(
                            StaticCredentialsProvider.create(AwsBasicCredentials.create("key", "secret")));
     }
@@ -58,7 +63,16 @@ public class GetBucketPolicyFunctionalTest {
     private S3AsyncClientBuilder getAsyncClientBuilder() {
         return S3AsyncClient.builder()
                             .region(Region.US_EAST_1)
-                            .endpointOverride(HTTP_LOCALHOST_URI)
+                            .endpointOverride(endpoint())
+                            .credentialsProvider(
+                                StaticCredentialsProvider.create(AwsBasicCredentials.create("key", "secret")));
+
+    }
+
+    private S3CrtAsyncClientBuilder getS3CrtAsyncClientBuilder() {
+        return S3AsyncClient.crtBuilder()
+                            .region(Region.US_EAST_1)
+                            .endpointOverride(endpoint())
                             .credentialsProvider(
                                 StaticCredentialsProvider.create(AwsBasicCredentials.create("key", "secret")));
 
@@ -82,5 +96,15 @@ public class GetBucketPolicyFunctionalTest {
 
         GetBucketPolicyResponse response = s3Client.getBucketPolicy(r -> r.bucket(EXAMPLE_BUCKET)).join();
         assertThat(response.policy()).isEqualTo(EXAMPLE_POLICY);
+    }
+
+    @Test
+    public void getBucketPolicy_crtClient() {
+        stubFor(any(anyUrl()).willReturn(aResponse().withStatus(200).withBody(EXAMPLE_POLICY)));
+
+        try(S3AsyncClient s3Client = getS3CrtAsyncClientBuilder().build()) {
+            GetBucketPolicyResponse response = s3Client.getBucketPolicy(r -> r.bucket(EXAMPLE_BUCKET)).join();
+            assertThat(response.policy()).isEqualTo(EXAMPLE_POLICY);
+        }
     }
 }

--- a/services/s3/src/test/java/software/amazon/awssdk/services/s3/internal/crt/S3CrtAsyncHttpClientTest.java
+++ b/services/s3/src/test/java/software/amazon/awssdk/services/s3/internal/crt/S3CrtAsyncHttpClientTest.java
@@ -27,8 +27,13 @@ import java.net.URI;
 import java.util.HashMap;
 import java.util.Map;
 import java.util.concurrent.CompletableFuture;
+import java.util.stream.Stream;
+import org.assertj.core.internal.Integers;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.MethodSource;
+import org.junit.jupiter.params.provider.ValueSource;
 import org.mockito.ArgumentCaptor;
 import org.mockito.Mockito;
 import software.amazon.awssdk.core.interceptor.trait.HttpChecksum;
@@ -42,6 +47,7 @@ import software.amazon.awssdk.http.SdkHttpRequest;
 import software.amazon.awssdk.http.async.AsyncExecuteRequest;
 import software.amazon.awssdk.http.async.SdkAsyncHttpResponseHandler;
 import software.amazon.awssdk.http.async.SdkHttpContentPublisher;
+import software.amazon.awssdk.services.s3.S3AsyncClient;
 
 public class S3CrtAsyncHttpClientTest {
     private static final URI DEFAULT_ENDPOINT = URI.create("https://127.0.0.1:443");
@@ -69,14 +75,23 @@ public class S3CrtAsyncHttpClientTest {
         asyncHttpClient = new S3CrtAsyncHttpClient(s3Client, s3NativeClientConfiguration);
     }
 
-    @Test
-    public void defaultRequest_shouldSetMetaRequestOptionsCorrectly() {
-        AsyncExecuteRequest asyncExecuteRequest = getExecuteRequestBuilder().build();
+    private static Stream<Integer> ports() {
+        return Stream.of(null, 1234, 443);
+    }
+
+    @ParameterizedTest
+    @MethodSource("ports")
+    public void defaultRequest_shouldSetMetaRequestOptionsCorrectly(Integer port) {
+        AsyncExecuteRequest asyncExecuteRequest = getExecuteRequestBuilder(port).build();
 
         S3MetaRequestOptions actual = makeRequest(asyncExecuteRequest);
         assertThat(actual.getMetaRequestType()).isEqualTo(S3MetaRequestOptions.MetaRequestType.DEFAULT);
         assertThat(actual.getCredentialsProvider()).isNull();
-        assertThat(actual.getEndpoint()).isEqualTo(URI.create(DEFAULT_ENDPOINT.getScheme() + "://" + DEFAULT_ENDPOINT.getHost()));
+        String expectedEndpoint = port == null || port.equals(443) ?
+                                  DEFAULT_ENDPOINT.getScheme() + "://" + DEFAULT_ENDPOINT.getHost() :
+                                  DEFAULT_ENDPOINT.getScheme() + "://" + DEFAULT_ENDPOINT.getHost() + ":" + port;
+        assertThat(actual.getEndpoint()).hasToString(expectedEndpoint);
+
 
         HttpRequest httpRequest = actual.getHttpRequest();
         assertThat(httpRequest.getEncodedPath()).isEqualTo("/key");
@@ -86,8 +101,9 @@ public class S3CrtAsyncHttpClientTest {
                                                  .collect(HashMap::new, (m, h) -> m.put(h.getName(), h.getValue())
                                                      , Map::putAll);
 
+        String expectedPort = port == null || port.equals(443)  ? "" : ":" + port;
         assertThat(headers).hasSize(4)
-                           .containsEntry("Host", DEFAULT_ENDPOINT.getHost())
+                           .containsEntry("Host", DEFAULT_ENDPOINT.getHost() + expectedPort)
                            .containsEntry("custom-header", "foobar")
                            .containsEntry("amz-sdk-invocation-id", "1234")
                            .containsEntry("Content-Length", "100");
@@ -293,6 +309,10 @@ public class S3CrtAsyncHttpClientTest {
     }
 
     private AsyncExecuteRequest.Builder getExecuteRequestBuilder() {
+        return getExecuteRequestBuilder(443);
+    }
+
+    private AsyncExecuteRequest.Builder getExecuteRequestBuilder(Integer port) {
         return AsyncExecuteRequest.builder()
                                   .responseHandler(responseHandler)
                                   .requestContentPublisher(contentPublisher)
@@ -300,7 +320,7 @@ public class S3CrtAsyncHttpClientTest {
                                                          .protocol(DEFAULT_ENDPOINT.getScheme())
                                                          .method(SdkHttpMethod.GET)
                                                          .host(DEFAULT_ENDPOINT.getHost())
-                                                         .port(DEFAULT_ENDPOINT.getPort())
+                                                         .port(port)
                                                          .encodedPath("/key")
                                                          .putHeader(CONTENT_LENGTH, "100")
                                                          .putHeader("amz-sdk-invocation-id",

